test/performance/odp_sched_pktio.c

```diff
@@ -0,0 +1,712 @@
+#include <stdio.h>
+#include <string.h>
+#include <signal.h>
+#include <stdlib.h>
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+
+#define MAX_WORKERS       64
+#define MAX_PKTIOS        32
+#define MAX_PKTIO_NAME    31
+#define MAX_PKTIO_QUEUES  MAX_WORKERS
+#define MAX_PKT_LEN       1514
+#define MAX_PKT_NUM       (16 * 1024)
+#define MIN_PKT_SEG_LEN   64
+#define BURST_SIZE        32
+#define CHECK_PERIOD      10000
+#define MAX_PKTIO_INDEXES 256
+
+typedef struct {
+	int  worker_id;
+	void *test_global_ptr;
+} worker_arg_t;
+
+typedef struct {
+	volatile int  stop_workers;
+	odp_barrier_t worker_start;
+
+	struct {
+		int num_worker;
+		int num_pktio;
+	} opt;
+
+	int max_workers;
+	odp_cpumask_t cpumask;
+	odp_instance_t instance;
+
+	int worker_cpu[MAX_WORKERS];
+
+	odp_pool_t pool;
+	uint32_t   pkt_len;
+	uint32_t   pkt_num;
+
+	unsigned int num_input_queues;
+	unsigned int num_output_queues;
+
+	struct {
+		char name[MAX_PKTIO_NAME + 1];
+		odp_pktio_t pktio;
+		int pktio_index;
+		int started;
+		odph_ethaddr_t my_addr;
+		odp_queue_t input_queue[MAX_PKTIO_QUEUES];
+		odp_pktout_queue_t pktout[MAX_PKTIO_QUEUES];
+
+	} pktio[MAX_PKTIOS];
+
+	worker_arg_t worker_arg[MAX_WORKERS];
+
+	/* Maps pktio input index to pktio[] index for output */
+	uint8_t pktio_map[MAX_PKTIO_INDEXES];
+
+} test_global_t;
+
+static test_global_t *test_global;
+
+static inline void set_dst_eth_addr(odph_ethaddr_t *eth_addr, int index)
+{
+	eth_addr->addr[0] = 0x02;
+	eth_addr->addr[1] = 0;
+	eth_addr->addr[2] = 0;
+	eth_addr->addr[3] = 0;
+	eth_addr->addr[4] = 0;
+	eth_addr->addr[5] = index;
+}
+
+static inline void fill_eth_addr(odp_packet_t pkt[], int num,
+				 test_global_t *test_global, int out)
+{
+	odph_ethhdr_t *eth;
+	int i;
+
+	for (i = 0; i < num; ++i) {
+		eth = odp_packet_data(pkt[i]);
+
+		eth->src = test_global->pktio[out].my_addr;
+		set_dst_eth_addr(&eth->dst, out);
+	}
+}
+
+static int worker_thread(void *arg)
+{
+	odp_event_t ev[BURST_SIZE];
+	int num, sent, drop, in, out;
+	odp_pktout_queue_t pktout;
+	worker_arg_t *worker_arg = arg;
+	test_global_t *test_global = worker_arg->test_global_ptr;
+	int worker_id = worker_arg->worker_id;
+	uint32_t polls = 0;
+
+	printf("Worker %i started\n", worker_id);
+
+	/* Wait for other workers to start */
+	odp_barrier_wait(&test_global->worker_start);
+
+	while (1) {
+		odp_packet_t pkt[BURST_SIZE];
+
+		num = odp_schedule_multi(NULL, ODP_SCHED_NO_WAIT,
+					 ev, BURST_SIZE);
+
+		polls++;
+
+		if (polls == CHECK_PERIOD) {
+			polls = 0;
+			if (test_global->stop_workers)
+				break;
+		}
+
+		if (num <= 0)
+			continue;
+
+		odp_packet_from_event_multi(pkt, ev, num);
+
+		in     = odp_packet_input_index(pkt[0]);
+		out    = test_global->pktio_map[in];
+		pktout = test_global->pktio[out].pktout[worker_id];
+
+		fill_eth_addr(pkt, num, test_global, out);
+
+		sent = odp_pktout_send(pktout, pkt, num);
+		drop = num - sent;
```

Comment:
Need to account for possibility that `sent < 0` here if `odp_pktout_send()` returned an error. Otherwise you'll calculate `drop > num` and attempt to reference a negative pkt index on the subsequent `odp_packet_free_multi()`.

Reply:
Done in v2.